src/lib/storage/subvol.rb

```diff
@@ -0,0 +1,171 @@
+#
+# To contact Novell about this file by physical or electronic mail, you may
+# find current contact information at www.novell.com.
+
+require "yast"
+require "storage"
+Yast.import "Arch"
+
+module Yast
+  class StorageClass < Module
+
+    # Helper class to represent a subvolume as defined in control.xml
+    #
+    class Subvol
+      include Yast::Logger
+
+      attr_accessor :path, :copy_on_write, :archs
+
+      COW_SUBVOL_PATHS = [
+        "home",
+        "opt",
+        "srv",
+        "tmp",
+        "usr/local",
+        "var/cache",
+        "var/crash",
+        "var/lib/machines",
+        "var/lib/mailman",
+        "var/lib/named",
+        "var/log",
+        "var/opt",
+        "var/spool",
+        "var/tmp"
+      ]
+
+      # No Copy On Write for SQL databases and libvirt virtual disks to
+      # minimize performance impact
+      NO_COW_SUBVOL_PATHS = [
+        "var/lib/libvirt/images",
+        "var/lib/mariadb",
+        "var/lib/mysql",
+        "var/lib/pgsql"
+      ]
+
+      def initialize(path, copy_on_write: true, archs: nil)
+        @path = path
+        @copy_on_write = copy_on_write
+        @archs = archs
+      end
+
+      def to_s
+        text = "Subvol #{@path}"
+        text += " (NoCOW)" unless @copy_on_write
+        text += " (archs: #{@archs})" if arch_specific?
+        text
+      end
+
+      def arch_specific?
+        !archs.nil?
+      end
+
+      def cow?
+        @copy_on_write
+      end
+
+      def no_cow?
+        !@copy_on_write
+      end
+
+      # Comparison operator for sorting
+      #
+      def <=>(other)
+        path <=> other.path
+      end
+
+      # Check if this subvolume should be used for the current architecture.
+      # A subvolume is used if its archs contain the current arch.
+      # It is not used if its archs contain the current arch negated
+      # (e.g. "!ppc").
+      #
+      # @return bool
+      #
+      def current_arch?
+        matches_arch? { |arch| Arch.respond_to?(arch.to_sym) && Arch.send(arch.to_sym) }
+      end
+
+      # Check if this subvolume should be used for an architecture.
+      #
+      # If a block is given, the block is called as the matcher with the
+      # architecture to be tested as its argument.
+      #
+      # If no block is given (and only then), the 'target_arch' parameter is
+      # used to check against.
+      #
+      def matches_arch?(target_arch = nil, &block)
+        return true unless arch_specific?
+        use_subvol = false
+        archs.each do |a|
+          arch = a.dup
+          negate = arch.start_with?("!")
+          arch[0] = "" if negate # remove leading "!"
+          if block_given?
+            match = block.call(arch)
+          else
+            match = arch == target_arch
+          end
+          if match && negate
+            log.info("Not using #{self} for explicitly excluded arch #{arch}")
+            return false
+          end
+          use_subvol ||= match
```

Comment:
Maybe you would like to use `archs.any?` instead of `each` so you don't need to take care of use_subvol. Moreover, if you find the architecture, the loop will end.